File: packages/sdk-py/src/arckit/clients/validation.py
"""ERC-8004 ValidationRegistry client."""

from __future__ import annotations

from web3 import Web3

from arckit._tx import TxContext
from arckit.abi import VALIDATION_REGISTRY_ABI
from arckit.types import ValidationResponse, ValidationStatus
from arckit.utils import hash_string


class ValidationClient:
    def __init__(self, ctx: TxContext, validation_registry_address: str):
        self._ctx = ctx
        self._contract = ctx.w3.eth.contract(
            address=Web3.to_checksum_address(validation_registry_address),
            abi=VALIDATION_REGISTRY_ABI,
        )

    def get_validation_status(self, request_hash: bytes) -> ValidationStatus:
        raw = self._contract.functions.getValidationStatus(request_hash).call()
        return ValidationStatus(
            validator_address=raw[0],
            agent_id=raw[1],
            response=ValidationResponse(raw[2]),
            response_hash=raw[3],
            tag=raw[4],
            last_update=raw[5],
        )

    def validation_request(
        self,
        validator: str,
        agent_id: int,
        request_uri: str,
        request_hash: bytes | None = None,
    ) -> dict:
        if request_hash is None:
            request_hash = hash_string(request_uri)
        return self._ctx.send(
            self._contract.functions.validationRequest(
                Web3.to_checksum_address(validator),
                int(agent_id),
                request_uri,
                request_hash,
            )
        )

    def validation_response(
        self,
        request_hash: bytes,
        response: ValidationResponse,
        response_uri: str,
        tag: str = "",
        response_hash: bytes | None = None,
    ) -> dict:
        if response_hash is None:
            response_hash = hash_string(response_uri)
        return self._ctx.send(
            self._contract.functions.validationResponse(
                request_hash,
                int(response),
                response_uri,
                response_hash,
                tag,
            )
        )

File: packages/sdk/src/clients/validation.ts
import type { Account, Address, Hex, PublicClient, WalletClient } from 'viem';
import { validationRegistryAbi } from '../abi/validationRegistry.js';
import { WalletRequiredError } from '../errors.js';
import type { ValidationResponse, ValidationStatus } from '../types.js';
import { hashString, waitForReceipt } from '../utils.js';

export type ValidationClientOpts = {
  publicClient: PublicClient;
  walletClient?: WalletClient;
  account?: Account;
  validationRegistry: Address;
};

export class ValidationClient {
  constructor(private readonly opts: ValidationClientOpts) {}

  // ── Reads ──

  async getValidationStatus(requestHash: Hex): Promise<ValidationStatus> {
    const raw = await this.opts.publicClient.readContract({
      address: this.opts.validationRegistry,
      abi: validationRegistryAbi,
      functionName: 'getValidationStatus',
      args: [requestHash],
    });
    return {
      validatorAddress: raw[0],
      agentId: raw[1],
      response: raw[2] as ValidationResponse,
      responseHash: raw[3],
      tag: raw[4],
      lastUpdate: raw[5],
    };
  }

  // ── Writes ──

  async validationRequest(params: {
    validator: Address;
    agentId: bigint | number;
    requestURI: string;
    requestHash?: Hex;
  }) {
    const { account, walletClient } = this.requireWallet('validationRequest');
    const requestHash = params.requestHash ?? hashString(params.requestURI);
    const hash = await walletClient.writeContract({
      account,
      chain: null,
      address: this.opts.validationRegistry,
      abi: validationRegistryAbi,
      functionName: 'validationRequest',
      args: [params.validator, BigInt(params.agentId), params.requestURI, requestHash],
    });
    return waitForReceipt(this.opts.publicClient, hash);
  }

  async validationResponse(params: {
    requestHash: Hex;
    response: ValidationResponse;
    responseURI: string;
    tag?: string;
    responseHash?: Hex;
  }) {
    const { account, walletClient } = this.requireWallet('validationResponse');
    const responseHash = params.responseHash ?? hashString(params.responseURI);
    const tag = params.tag ?? '';
    const hash = await walletClient.writeContract({
      account,
      chain: null,
      address: this.opts.validationRegistry,
      abi: validationRegistryAbi,
      functionName: 'validationResponse',
      args: [params.requestHash, params.response, params.responseURI, responseHash, tag],
    });
    return waitForReceipt(this.opts.publicClient, hash);
  }

  private requireWallet(op: string): { account: Account; walletClient: WalletClient } {
    if (!this.opts.account || !this.opts.walletClient) {
      throw new WalletRequiredError(op);
    }
    return { account: this.opts.account, walletClient: this.opts.walletClient };
  }
}
